weasyl/controllers/user.py

```diff
@@ -42,6 +43,21 @@ def signin_post_(request):
         # Invalidate cached versions of the frontpage to respect the possibly changed SFW settings.
         index.template_fields.invalidate(logid)
         raise HTTPSeeOther(location=form.referer)
+    elif logid and logerror == "2fa":
+        # Password authentication passed, but user has 2FA set, so verify second factor (Also set SFW mode now)
+        if form.sfwmode == "sfw":
+            request.set_cookie_on_response("sfwmode", "sfw", 31536000)
+        index.template_fields.invalidate(logid)
+        # Store the authenticated userid & password auth time to the session
```

Comment:
This may be me being insanely careful about this, but maybe check if a zero-recovery code situation exists here? In theory it should *never* happen (we catch the zero situation down in ``signin_2fa_auth_post_(request)``); but this is me considering a situation where some database lock will mess someone up, even though yes, I know PSQL actions are atomic. Call me paranoid.

Anyway, pseudocode:
```
[[[ Here, in weasyl/controllers/user.py ]]]
remaining_recovery_codes = two_factor_auth.get_number_of_recovery_codes(logid)
if remaining_recovery_codes == 0:
    # The user's account (inexplicably) has zero recovery codes remaining. Force-disable 2FA.
    two_factor_auth.force_disable(logid)
    raise WeasylError("TwoFactorAuthentication-ZeroRecoveryCodesRemaining")  # Note to self, why do error codes not seem to use dashes?

[[[ Over in weasyl/errorcode.py ]]]
"TwoFactorAuthentication-ZeroRecoveryCodesRemaining": "Your account has zero recovery codes remaining, and has had 2FA disabled to prevent you from being permanently unable to log into your account. You may re-enable 2FA if you desire to do so."

[[[ Down in signin_2fa_auth_post_(request) where the zero code check happens ]]]
Do the same thing. Raise the error versus using the custom message?
```

Just my thoughts for now.